03_Predicting_Inverse_Temperature.ipynb

```diff
@@ -0,0 +1,1377 @@
+        "\n",
+        "We are also going to try and run the simulation \"backwards\".  In the previous notebook, a value for the inverse temperature was chosen first, and then we simulated a lattice based on the probability of that lattice for the given inverse temperature value.  Here, we are going to start with a lattice where the temperature is unknown, then work backwards to see if we can figure out what inverse temperature might have allowed this particular lattice to form.   We will try to figure it out based on all of the other lattice-temperature relationships. "
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "qJAmGFSmOoIB"
+      },
+      "source": [
+        "# Exercises\n",
+        "\n",
+        "The first code blocks in the `Generate data for usage` section can be executed directly to generate the training and test data for the coding exercises later.\n",
+        "\n",
+        "The programming exercises here are meant to help you gain intuition about how the different kinds of models behave, so you will use pre-defined functions and models when they are available.  If you would like to try coding some models from scratch, there are many tutorials on the internet that are beyond the scope of this notebook.\n",
+        "\n",
+        "**Programming Exercise 1** walks you through implementing a *Support Vector Machine - Regression* or SVR. \n",
+        "\n",
+        "**Programming Exercise 2** walks you through implementing a *Fully-Connected Deep Neural Network* or FC-DNN.\n",
+        "\n",
+        "**Programming Exercise 3** walks you through implementing a *2D Convolutional Neural Network* or CNN."
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "wmiC_qfZOqmm"
+      },
+      "source": [
+        "## Generate data for usage\n",
+        "\n",
+        "Before training the models, we need need data to train them with.  Executing the following code cell will create the functions needed to duplicate the simulations from the previous notebook."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "7ejdDQlrO2LD"
+      },
+      "outputs": [],
+      "source": [
+        "def initialize_lattice(L: int, c1 = 0):\n",
+        "    \"\"\"    \n",
+        "    Function to initialize lattice.  Adds a border of zeros\n",
+        "    to represent non-interacting atoms and make the neighbor\n",
+        "    calculation easier\n",
+        "\n",
+        "    L: The square root of the number of atoms in the lattice\n",
+        "    returns padded_lattice: A lattice of size (L+2, L+2)\n",
+        "    \"\"\"\n",
+        "\n",
+        "    # initialize the lattice using np.random.random for a lattice of size\n",
+        "    # LxL\n",
+        "\n",
+        "    init_lattice = np.random.random(size=(L,L))\n",
```

Comment:
I mentioned this in notebook 2 but it may be easier to explain if this function is rewritten along the lines of
```
init_lattice = np.random.choice([-1,1], size=(L,L))
padded_lattice = np.zeros((L+2,L+2))
padded_lattice[1:-1,1:-1] = init_lattice
return padded_lattice
```

Reply:
Thank you for your comment.  Since I am explicitly reusing the code from N2, I do not want to introduce changes here that are not in N2.  See comment above for why I am not changing the main algorithm in N2 to use this function.